ethereum/contracts/Bridge.sol

```diff
@@ -0,0 +1,61 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.6.2;
+
+import "./Decoder.sol";
+import "./Application.sol";
+
+contract Bridge {
+    using Decoder for bytes;
+
+    // 0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470 is returned for accounts without code
+    bytes32 NON_CONTRACT_ACCOUNT_HASH = 0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470;
+
+    mapping(bytes32 => bool) public eventHashes;
+    mapping(address => bool) public applications;
+
+    constructor(address[] memory apps) public {
+        for(uint256 i = 0; i < apps.length; i++) {
+            if(verifyApp(apps[i])) {
+                applications[apps[i]] = true;
+            }
+        }
+    }
+
+    /**
+     * @dev routes the message to the specified application ID after verifying the operator's signature
+     * @param _message address _message expected type: Message { AppID [32]byte, Payload []byte }
+     * @param _hash bytes _hash is a unique hash of the event (app_id, payload, blockNum, eventId)
+     */
+    function submit(bytes memory _message, bytes32 _hash)
+        public
+    {
+        address appID = _message.sliceAddress(32);
+        require(applications[appID], "App ID not found. Only registered application are supported");
+
+        require(!eventHashes[_hash], "Event has already been processed");
+        eventHashes[_hash] = true;
+
+        bytes memory payload = _message.slice(32, _message.length-1);
+        Application app = Application(appID);
+        app.handle(payload);
```

Comment:
Don't we have to call into the verifier contract to verify that the sender's address is the one we trust.

For reference, here is the main entry point into the verifier on the Substrate side:
https://github.com/Snowfork/polkadot-ethereum/blob/app-identifiers/parachain/pallets/verifier/src/lib.rs#L52

Reply:
So along those lines I'd suggest adding a method `Verifier.verify(address _sender, bytes _message)` to the verifier, have the bridge call that.

Reply:
The Verifier contract expects the signature to be passed separately from the contents of the signed message.
1. `verify(bytes32 _hash, bytes memory _signature)`
2. `verifyBytes(bytes memory _rawMessage, bytes memory _signature)`

So we need to parse the signature from the `_message`. But in my understanding, the message currently doesn't contain a signature?

Side note: I think some of the confusion around naming conventions here stems from the fact that Ethereum generally uses "message" to refer to the data that an address applied their signature to.

Reply:
The main branch currently has the verification data as: 
```
type VerificationBasic struct {
	BlockNumber uint64
	EventIndex  uint32
}
```

Is the proposal here that we should
1. Parse `Message.VerificationBasic.BlockNumber` and `Message.VerificationBasic.EventIndex`
2. Create a hash via `hash(AppID, BlockNumber, EventIndex, Message.Data)`
3. Validate that we haven't processed this hash yet
4. Check Verifier contract to see if the signature is from the operator

I'm confused because there is no signature available, we've replaced it with the BlockNumber/EventIndex, correct?

Reply:
Or the Verifier contract just checks msg.sender and doesn't check the signature?

Reply:
Just checking msg.sender and performing a hash of the entire message should be sufficient:

Basically this rough pseudocode shows what the bridge needs to do:
```
submit(app_id, message):
  verifySender(msg.sender)
  VerifyHash(hash(app_id, message))
```

Since the block number and event index are embedded in the message bytes, we can just hash the entire message for replay protection

Right now the relayer isn't sending the signature anymore. Also type `types.VerificationBasic` in the relayer isn't being used anymore.

Reply:
updated in https://github.com/Snowfork/polkadot-ethereum/pull/75/commits/0f905430101dc05b1624c155753ac5169d4f6624, https://github.com/Snowfork/polkadot-ethereum/pull/75/commits/c07a362125122d0bfc0f8030e84165872d3d9dd4